File: elixir/luhn/luhn.exs
defmodule Luhn do
  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    number |> parse |> check
  end

  defp parse(number) do
    if String.match?(number, ~r/[^\d|\s]/) do
      []
    else
      number
      |> String.graphemes()
      |> Enum.reject(&(&1 == " "))
      |> Enum.map(&String.to_integer/1)
      |> Enum.reverse()
    end
  end

  defp check(numbers) when length(numbers) <= 1 do
    false
  end

  defp check(numbers) do
    numbers
    |> Enum.with_index(1)
    |> Enum.flat_map(&calc/1)
    |> Enum.sum()
    |> rem(10) == 0
  end

  defp calc({n, i}) when rem(i, 2) != 0, do: [n]

  defp calc({n, _}) do
    res = n * 2

    if res >= 10 do
      [1, rem(res, 10)]
    else
      [res]
    end
  end
end

File: typescript/luhn/luhn.ts
export default class Luhn {
    static valid(input: string): boolean {
        if (!/\d|\s/.test(input)) return false
        if (input.trim().length <= 1) return false

        const sum = input.replace(/\s/g, '').split('')
            .reduceRight((acc, val, idx, arr) => acc + this.calc(Number(val), arr.length - idx), 0)
        return sum % 10 === 0
    }

    static calc(n: number, rnum: number): number {
        if (rnum % 2 === 0) {
            const res = n * 2
            return (res >= 9) ? res - 9 : res
        }
        return n
    }
}